### PR TITLE
Issue 1249: Migration - GIS

### PR DESCRIPTION
**Pull request by dleszcz**
#1249 

### Description

- Created a new panel component
- Created a new GIS component
- Created a GIS module to use ViewerService
- Removed Angular GIS component
- Implement styles with styled-components

**Bugs**

- [x] Technically the Drop down menu is like a heading and the Map Tiles etc list are within that heading. So it looks a bit weird how the text on the drop down menu is a lot more insignificant to the Map Tiles etc. Can we format this a bit better? maybe bigger font on the drop down with black text and make the list text smaller + a dark grey? (whatever you feel appropriate)

- [x] Capitalise Base here:

- [x] The settings page is still showing "mm" even though my model is in "m"

- [x] Coordinates needs swapping. Due to us working in openGL coordinates and CAD users use direct x + they want z to have a negated meaning... [X Y Z] on the UI needs to be sent as [X -Z -Y] to the backend.

- [x] This is GIS branch (can we make sure the settings page also has this logic please :) ):

- [x] Save button needs inverting (should be a round green button with white floppy disk):

### CODE REVIEW

File: frontend/routes/viewer/components/viewerCard/viewerCard.component.tsx [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+		return (
+			<ViewerCardTitle
+				title={title}
+				Icon={Icon}
+				renderActions={() => this.renderTitleActions(actions)}
+			/>
+		);
+	}
+
+	public renderTitleActions = (actions) => (
```
[the remark this answers is not in the document]
Comment (carmenfan) [reply frontend/routes/viewer/components/viewerCard/viewerCard.component.tsx]: Can we call this `panel` like the angular one? viewerCard makes sense, but it'll be easier for guys who already know the angular code base to know what this is.

[the remark this answers is not in the document]
Comment (carmenfan) [reply frontend/routes/viewer/components/viewerCard/viewerCard.component.tsx]: yes I think ViewerPanel would be great :)

File: frontend/routes/viewer/components/gis/components/settingsForm/settingsForm.component.tsx [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+	public componentDidMount = () => {
+		const changes = {} as any;
+
+		if (!isEmpty(this.props.initialValues)) {
+			const { latitude, longitude, axisX, axisY, axisZ, angleFromNorth } = this.state;
+			const {
+				latitude: lat, longitude: lng, axisX: x, axisY: y, axisZ: z, angleFromNorth: afn
+			} = this.props.initialValues;
+
+			if (!isEqual(latitude, lat)) {
```
[the remark this answers is not in the document]
Comment (dleszcz) [reply frontend/routes/viewer/components/gis/components/settingsForm/settingsForm.component.tsx]: Updated form implementation using `withFormik`. 
Also corrected your other comments.

File: frontend/routes/viewer/components/viewerCard/viewerCard.component.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (dleszcz) [reply frontend/routes/viewer/components/viewerCard/viewerCard.component.tsx]: ok, so i renamed this component

**Comment by simplypixi** [comment]
@carmenfan Rebase and ready for review ;)

File: frontend/routes/viewer/components/gis/gis.component.tsx [hunk 3]
```diff
@@ -0,0 +1,289 @@
...
+		const { settingsModeActive } = this.state;
+
+		return (
+			<ViewerPanel
+				title="GIS"
+				Icon={this.getTitleIcon()}
+				actions={this.getActions()}
+				pending={this.props.isPending}
+			>
+				{settingsModeActive ? this.renderSettings() : this.renderMapLayers()}
```
Comment (sanmont3drepo) [inline frontend/routes/viewer/components/gis/gis.component.tsx]: I'd separate renderSettings() and renderMapLayers() into to differents components/files.

Comment (dleszcz) [reply frontend/routes/viewer/components/gis/gis.component.tsx]: I moved Settings to separate component, but referring to the layers, we think that it's so small that we don't need to divide it into smaller ones

**Comment by simplypixi** [comment]
@carmenfan Fixed and rebased ;)

**Comment by simplypixi** [comment]
@carmenfan 

> The panel needs to resize depending on the height available + have the ability to scroll

I've fixed this but it was a big challenge. I created the new extra angular component to plug into your algorithm. It's not the most beautiful solution but is the only one, which can we provide in this situation.

Moreover, while fixing I noticed a few extra bugs:
- the panels re-render too many times
- while website initializing each panel is visible for a short time (even if it's closed)
- regardless of GIS migration, the panels still can have too big height (e.g opening more than three panels)

Please, tell me what you think about it. ;)

**Comment by carmenfan** [comment]
> 
> 
> @carmenfan
> 
> > The panel needs to resize depending on the height available + have the ability to scroll
> 
> I've fixed this but it was a big challenge. I created the new extra angular component to plug into your algorithm. It's not the most beautiful solution but is the only one, which can we provide in this situation.
> 
> Moreover, while fixing I noticed a few extra bugs:
> 
>     * the panels re-render too many times
> 
>     * while website initializing each panel is visible for a short time (even if it's closed)
> 
>     * regardless of GIS migration, the panels still can have too big height (e.g opening more than three panels)
> 
> 
> Please, tell me what you think about it. ;)

How many of those issues are angular existing code which will go away when you migrate them to react? If it's a temporary thing we can probably ignore it.

**Comment by simplypixi** [comment]
> > @carmenfan
> > > The panel needs to resize depending on the height available + have the ability to scroll
> > 
> > 
> > I've fixed this but it was a big challenge. I created the new extra angular component to plug into your algorithm. It's not the most beautiful solution but is the only one, which can we provide in this situation.
> > Moreover, while fixing I noticed a few extra bugs:
> > ```
> > * the panels re-render too many times
> > 
> > * while website initializing each panel is visible for a short time (even if it's closed)
> > 
> > * regardless of GIS migration, the panels still can have too big height (e.g opening more than three panels)
> > ```
> > Please, tell me what you think about it. ;)
> 
> How many of those issues are angular existing code which will go away when you migrate them to react? If it's a temporary thing we can probably ignore it.

All of them are directly connected with angular code. So everything will be fine after migration. :)

**Comment by simplypixi** [comment]
@carmenfan Could you review this PR?

**Comment by carmenfan** [comment]
@simplypixi it seems to reload everytime I open/close the panel? I think we're ok with it not fetching the data everytime we show/hide and just keep the last state, if that's what it's doing.

before


Everything seems to work otherwise! :)

**Comment by carmenfan** [comment]
@simplypixi 

I found a weird one. So far I only have problem with this model so I'm struggling to find the pattern...

When I try to load model settings (both on this page and on the GIS tab), it gets stuck and there's an error on the console:
image
image

Model settings looks like this on database:

```json
{
    "_id" : "067a7522-6dd8-4258-97a0-be756e596324",
    "timestamp" : ISODate("2017-10-30T12:08:18.000Z"),
    "status" : "ok",
    "type" : "Structural",
    "desc" : "",
    "owner" : "carmen",
    "name" : "Clinic_S",
    "subModels" : [],
    "properties" : {
        "code" : "STR",
        "unit" : "mm",
        "topicTypes" : [ 
            {
                "value" : "for_information",
                "label" : "For information"
            }, 
            {
                "value" : "vr",
                "label" : "VR"
            }, 
            {
                "value" : "clash",
                "label" : "Clash"
            }, 
            {
                "value" : "diff",
                "label" : "Diff"
            }, 
            {
                "value" : "rfi",
                "label" : "RFI"
            }, 
            {
                "value" : "risk",
                "label" : "Risk"
            }, 
            {
                "value" : "hs",
                "label" : "H&S"
            }, 
            {
                "value" : "design",
                "label" : "Design"
            }, 
            {
                "value" : "constructibility",
                "label" : "Constructibility"
            }, 
            {
                "value" : "gis",
                "label" : "GIS"
            }
        ]
    },
    "permissions" : [ 
        {
            "user" : "carmen",
            "permission" : "collaborator"
        }
    ],
    "users" : [],
    "__v" : 12,
    "surveyPoints" : [],
    "heliSpeed" : -8
}
````

Works on staging so it seems to be related to something that was updated on this branch. Any ideas?

**Comment by simplypixi** [comment]
> @simplypixi
> 
> I found a weird one. So far I only have problem with this model so I'm struggling to find the pattern...
> 
> When I try to load model settings (both on this page and on the GIS tab), it gets stuck and there's an error on the console:
> image
> image
> 
> Model settings looks like this on database:
> 
> ```json
> {
>     "_id" : "067a7522-6dd8-4258-97a0-be756e596324",
>     "timestamp" : ISODate("2017-10-30T12:08:18.000Z"),
>     "status" : "ok",
>     "type" : "Structural",
>     "desc" : "",
>     "owner" : "carmen",
>     "name" : "Clinic_S",
>     "subModels" : [],
>     "properties" : {
>         "code" : "STR",
>         "unit" : "mm",
>         "topicTypes" : [ 
>             {
>                 "value" : "for_information",
>                 "label" : "For information"
>             }, 
>             {
>                 "value" : "vr",
>                 "label" : "VR"
>             }, 
>             {
>                 "value" : "clash",
>                 "label" : "Clash"
>             }, 
>             {
>                 "value" : "diff",
>                 "label" : "Diff"
>             }, 
>             {
>                 "value" : "rfi",
>                 "label" : "RFI"
>             }, 
>             {
>                 "value" : "risk",
>                 "label" : "Risk"
>             }, 
>             {
>                 "value" : "hs",
>                 "label" : "H&S"
>             }, 
>             {
>                 "value" : "design",
>                 "label" : "Design"
>             }, 
>             {
>                 "value" : "constructibility",
>                 "label" : "Constructibility"
>             }, 
>             {
>                 "value" : "gis",
>                 "label" : "GIS"
>             }
>         ]
>     },
>     "permissions" : [ 
>         {
>             "user" : "carmen",
>             "permission" : "collaborator"
>         }
>     ],
>     "users" : [],
>     "__v" : 12,
>     "surveyPoints" : [],
>     "heliSpeed" : -8
> }
> ```
> 
> Works on staging so it seems to be related to something that was updated on this branch. Any ideas?

The error, which is visible in dev tools, appears because of another error, which cannot be handled by error dialog. We have to investigate that but unfortunately, we can do it until after the Christmas break... :(

**Comment by carmenfan** [comment]
@simplypixi Yep returns 200 and something that looks sane:
image


We can revisit in the new year ;) I might try more models to see how wide spread this is and if not then I will merge this and put it up as a separate bug

**Comment by simplypixi** [comment]
> > @simplypixi it seems to reload everytime I open/close the panel? I think we're ok with it not fetching the data everytime we show/hide and just keep the last state, if that's what it's doing.
> > before
> > Everything seems to work otherwise! :)
> 
> Yes, it reloads every time. We've implemented this behavior before the discussion about saving panel state.
> I'll change that ;)

@carmenfan I almost finished this one, but I have a few questions.
What should happen, if the user sets "Map Tile" as visible and then closes GIS panel? This layer should be still visible?

**Comment by carmenfan** [comment]
> > > @simplypixi it seems to reload everytime I open/close the panel? I think we're ok with it not fetching the data everytime we show/hide and just keep the last state, if that's what it's doing.
> > > before
> > > Everything seems to work otherwise! :)
> > 
> > 
> > Yes, it reloads every time. We've implemented this behavior before the discussion about saving panel state.
> > I'll change that ;)
> 
> @carmenfan I almost finished this one, but I have a few questions.
> What should happen, if the user sets "Map Tile" as visible and then closes GIS panel? Should be this layer still visible?

Yes. It should still be visible :)

**Comment by simplypixi** [comment]
> > > > @simplypixi it seems to reload everytime I open/close the panel? I think we're ok with it not fetching the data everytime we show/hide and just keep the last state, if that's what it's doing.
> > > > before
> > > > Everything seems to work otherwise! :)
> > > 
> > > 
> > > Yes, it reloads every time. We've implemented this behavior before the discussion about saving panel state.
> > > I'll change that ;)
> > 
> > 
> > @carmenfan I almost finished this one, but I have a few questions.
> > What should happen, if the user sets "Map Tile" as visible and then closes GIS panel? Should be this layer still visible?
> 
> Yes. It should still be visible :)

Fixed ;)

**Comment by simplypixi** [comment]
@carmenfan Could you check this PR?

**Comment by simplypixi** [comment]
@carmenfan Fixed ;)

**Pull request merged**